pkg/piecestore/psserver/agreementsender/agreementsender.go

```diff
@@ -10,121 +10,96 @@ import (
+	return &AgreementSender{DB: DB, log: log, transport: transport.NewClient(identity), kad: kad}
 }
 
-// Run the afreement sender with a context to cehck for cancel
-func (as *AgreementSender) Run(ctx context.Context) error {
-	zap.S().Info("AgreementSender is starting up")
-
-	type agreementGroup struct {
-		satellite  storj.NodeID
-		agreements []*psdb.Agreement
-	}
-
-	c := make(chan *agreementGroup, 1)
-
+// Run the afreement sender with a context to check for cancel
+func (as *AgreementSender) Run(ctx context.Context) {
+	//todo:  we likely don't want to stop on err, but consider returning errors via a channel
 	ticker := time.NewTicker(*defaultCheckInterval)
 	defer ticker.Stop()
-	go func() {
-		for range ticker.C {
-			agreementGroups, err := as.DB.GetBandwidthAllocations()
-			if err != nil {
-				zap.S().Error(err)
-				continue
-			}
-
-			// Send agreements in groups by satellite id to open less connections
-			for satellite, agreements := range agreementGroups {
-				c <- &agreementGroup{satellite, agreements}
-			}
-		}
-	}()
-
 	for {
+		as.log.Debug("AgreementSender is running", zap.Duration("duration", *defaultCheckInterval))
+		agreementGroups, err := as.DB.GetBandwidthAllocations()
+		if err != nil {
+			as.log.Error("Agreementsender could not retrieve bandwidth allocations", zap.Error(err))
+			continue
+		}
+		for satellite, agreements := range agreementGroups {
+			as.sendAgreementsToSatellite(ctx, satellite, agreements)
+		}
 		select {
+		case <-ticker.C:
 		case <-ctx.Done():
-			return utils.CombineErrors(as.errs...)
-		case agreementGroup := <-c:
-			go func() {
-				zap.S().Infof("Sending %v agreements to satellite %s\n", len(agreementGroup.agreements), agreementGroup.satellite)
-
-				// Get satellite ip from overlay by Lookup agreementGroup.satellite
-				satellite, err := as.overlay.Lookup(ctx, agreementGroup.satellite)
-				if err != nil {
-					zap.S().Error(err)
-					return
-				}
-
-				// Create client from satellite ip
-				identOpt, err := as.identity.DialOption(storj.NodeID{})
-				if err != nil {
-					zap.S().Error(err)
-					return
-				}
-
-				conn, err := grpc.Dial(satellite.GetAddress().Address, identOpt)
-				if err != nil {
-					zap.S().Error(err)
-					return
-				}
-
-				client := pb.NewBandwidthClient(conn)
-
-				for _, agreement := range agreementGroup.agreements {
-
-					msg := &pb.RenterBandwidthAllocation{
-						Data:      agreement.Agreement,
-						Signature: agreement.Signature,
-					}
+			as.log.Debug("AgreementSender is shutting down", zap.Error(ctx.Err()))
+			return
+		}
+	}
+}
 
-					// Send agreement to satellite
-					r, err := client.BandwidthAgreements(ctx, msg)
-					if err != nil || r.GetStatus() != pb.AgreementsSummary_OK {
-						zap.S().Errorf("Failed to send agreement to satellite: %+v", err)
-						return
-					}
+func (as *AgreementSender) sendAgreementsToSatellite(ctx context.Context, satID storj.NodeID, agreements []*psdb.Agreement) {
```

Comment:
I'm wondering if wouldn't be more cleaner to add error to function response and use `ASError` to wrap things but not 100% sure.

Reply:
I just tried refactoring to that, and it traded a few simple lines of repeated code for added complexity.  I decided to leave it for fear of introducing new bugs.

Reply:
:+1: